### PR TITLE
docs(homepage): auto-update Layer ARN on every release

### LINKED ISSUE #1171: Publish the Public Layer ARN  for aws-lambda-powertools-python-layer-extras in the docs

### LINKED ISSUE #1549: Maintenance: create mechanism to auto publish the layer version for every release
### Summary

Right now, everytime we craft a new release, we have to wait for the Layer Pipeline to finish and manually update/re-generate the documentation to update the Layer ARN version.

We should automate this as part of our release pipeline.

### Why is this needed?

In order to reduce maintenance overhead.

### Which area does this relate to?

Automation

### Solution

The CDK Layer Pipeline already outputs the LayerVersionArn for each region. We can leverage this to automate the replacement of the version in our docs.

### Acknowledgment

- [X] This request meets [Lambda Powertools Tenets](https://awslabs.github.io/aws-lambda-powertools-python/latest/#tenets)
- [ ] Should this be considered in other Lambda Powertools languages? i.e. [Java](https://github.com/awslabs/aws-lambda-powertools-java/), [TypeScript](https://github.com/awslabs/aws-lambda-powertools-typescript/)

### PR DESCRIPTION
**Pull request by rubenfonseca**
**Issue number:** #1549

## Summary

This PR automates the publishing of the layer version in our documetation for every release.
This will reduce maintance overhead, and make the publishing of the new layer version faster.

### Changes

> Please provide a summary of what's being changed

- Update action-lint to support multiple nested workflow calls https://github.blog/changelog/2022-08-22-github-actions-improvements-to-reusable-workflows-2/
- Added arm64 layers to docs
- Added method to republish docs after a new published Layer

### User experience

> Please share what the user experience looks like before and after this change

After this change, the new layer version will be update don the documentation automatically, as soon as the Layer is created and published.

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

-----
[View rendered docs/index.md](https://github.com/rubenfonseca/aws-lambda-powertools-python/blob/feat/publish-layer/docs/index.md)

### CODE REVIEW

File: .github/workflows/reusable_publish_v2_layer_arn.yml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        with:
+          name: cdk-layer-stack
+          path: cdk-layer-stack
+      - name: Replace layer versions in documentation
+        run: ./layer/scripts/update_layer_arn.sh cdk-layer-stack
+      - name: Update documentation in trunk
+        run: |
+          HAS_CHANGE=$(git status --porcelain)
+          test -z "${HAS_CHANGE}" && echo "Nothing to update" && exit 0
+          git add CHANGELOG.md
```
Comment (heitorlessa) [inline .github/workflows/reusable_publish_v2_layer_arn.yml]: docs?

File: .github/workflows/reusable_publish_v2_layer_arn.yml [hunk 2]
```diff
@@ -0,0 +1,60 @@
+name: Publish V2 Layer ARN
```
Comment (heitorlessa) [inline .github/workflows/reusable_publish_v2_layer_arn.yml]: Maybe add `docs` in the same so to not be confused with the actual deploy/publish layer

File: layer/scripts/update_layer_arn.sh [hunk 3]
```diff
@@ -0,0 +1,59 @@
+#!/bin/bash
+
+# This script is run during the reusable_publish_v2_layer_arn CI job,
```
Comment (heitorlessa) [inline layer/scripts/update_layer_arn.sh]: *docs

File: layer/scripts/update_layer_arn.sh [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+# and it is responsible for replacing the layer ARN in our documentation,
+# based on the output files generated by CDK when deploying to each region.
+#
+# see .github/workflows/reusable_deploy_v2_layer_stack.yml
+
```
Comment (heitorlessa) [inline layer/scripts/update_layer_arn.sh:8]: pr: add more comments on expected input, expected transformation, etc.

File: layer/scripts/update_layer_arn.sh [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    # arn:aws:lambda:{env.region}:017000801446:layer:AWSLambdaPowertoolsPython:39
+    if [[ "$line" == *"eu-central-1"* ]]; then
+      for region in '{region}' '${AWS::Region}' '${aws:region}' '{env.region}'
+      do
+        generic_prefix=$(echo "$prefix" | sed "s/eu-central-1/${region}/")
+        generic_line=$(echo "$line" | sed "s/eu-central-1/${region}/")
+
+        sed -i '' -e "s/$generic_prefix:[[:digit:]][[:digit:]]*/$generic_line/g" docs/index.md
+      done
+    fi
```
Comment (heitorlessa) [inline layer/scripts/update_layer_arn.sh]: same as before - more comments on expected input, expected transformation, etc.

might be easier to reference the "canonical" as the Lambda Layers example in various Frameworks (e.g., SAM, SLS, etc.)

and that `eu-central-1` is simply a region that you're picking to get the version and replace each potential **region variation** of a line

File: layer/scripts/update_layer_arn.sh [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    # Sed doesn't support \d+ in a portable way, so we cheat with (:digit: :digit: *)
+    sed -i '' -e "s/$prefix:[[:digit:]][[:digit:]]*/$line/g" docs/index.md
+
+    # We use the eu-central-1 layer as the canonical version for all the generic examples
+    # arn:aws:lambda:{region}:017000801446:layer:AWSLambdaPowertoolsPython:39
+    # arn:aws:lambda:${AWS::Region}:017000801446:layer:AWSLambdaPowertoolsPython:39
+    # arn:aws:lambda:${aws:region}:017000801446:layer:AWSLambdaPowertoolsPython:39
+    # arn:aws:lambda:{env.region}:017000801446:layer:AWSLambdaPowertoolsPython:39
+    if [[ "$line" == *"eu-central-1"* ]]; then
+      for region in '{region}' '${AWS::Region}' '${aws:region}' '{env.region}'
```
Comment (heitorlessa) [inline layer/scripts/update_layer_arn.sh]:
```suggestion
      for pseudo_region in '{region}' '${AWS::Region}' '${aws:region}' '{env.region}'
```

pseudo_region = AWS::Region, etc.

File: .github/workflows/reusable_publish_v2_layer_arn.yml [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+  release-docs:
+    needs: publish_v2_layer_arn
+    permissions:
+      contents: write
+      pages: write
+    uses: ./.github/workflows/reusable_publish_docs.yml
+    with:
+      version: ${{ inputs.latest_published_version }}
+      alias: alpha
```
Comment (heitorlessa) [inline .github/workflows/reusable_publish_v2_layer_arn.yml]:
```suggestion
      alias: latest
```

**Review by heitorlessa — CHANGES_REQUESTED** [review]
two minor issues (git add missing docs, and docs alias) - and one major to add comments on shell script as not everyone is a seasoned bash (and regex) wizardry like you ;D

**Comment by rubenfonseca** [comment]
Note: update the rest of the SAM sample code on docs/index.md

**Comment by rubenfonseca** [comment]
Note: remove "Can't find our Lambda Layer for your preferred AWS region?" because we are now publishing in all regions.

**Comment by heitorlessa** [comment]
Docs looks excellent. Perhaps let's just increase md-grid (where the content is) to 90vw to give a bit more room, otherwise the layer per region table will look too tight

File: layer/scripts/update_layer_arn.sh [hunk 8]
```diff
@@ -36,22 +36,38 @@ do
...
     # arn:aws:lambda:${aws:region}:017000801446:layer:AWSLambdaPowertoolsPython:39
     # arn:aws:lambda:{env.region}:017000801446:layer:AWSLambdaPowertoolsPython:39
     if [[ "$line" == *"eu-central-1"* ]]; then
-      for region in '{region}' '${AWS::Region}' '${aws:region}' '{env.region}'
+      # These are all the framework pseudo parameters currently found in the docs
+      for pseudo_region in '{region}' '${AWS::Region}' '${aws:region}' '{env.region}'
       do
-        generic_prefix=$(echo "$prefix" | sed "s/eu-central-1/${region}/")
-        generic_line=$(echo "$line" | sed "s/eu-central-1/${region}/")
+        generic_prefix=$(echo "$prefix" | sed "s/eu-central-1/${pseudo_region}/")
```
Comment (heitorlessa) [inline layer/scripts/update_layer_arn.sh]:
```suggestion
        prefix_pseudo_region=$(echo "$prefix" | sed "s/eu-central-1/${pseudo_region}/")
```

File: layer/scripts/update_layer_arn.sh [hunk 9]
```diff
@@ -36,22 +36,38 @@ do
...
+        # generic_prefix = arn:aws:lambda:${AWS::Region}:017000801446:layer:AWSLambdaPowertoolsPython
+
+        generic_line=$(echo "$line" | sed "s/eu-central-1/${pseudo_region}/")
```
Comment (heitorlessa) [inline layer/scripts/update_layer_arn.sh]:
```suggestion
        line_pseudo_region=$(echo "$line" | sed "s/eu-central-1/${pseudo_region}/")
```

File: layer/scripts/update_layer_arn.sh [hunk 10]
```diff
@@ -36,22 +36,38 @@ do
...
+        # generic_line = arn:aws:lambda:${AWS::Region}:017000801446:layer:AWSLambdaPowertoolsPython:49
```
Comment (heitorlessa) [inline layer/scripts/update_layer_arn.sh]:
```suggestion
        # line_pseudo_region = arn:aws:lambda:${AWS::Region}:017000801446:layer:AWSLambdaPowertoolsPython:49
```

File: layer/scripts/update_layer_arn.sh [hunk 11]
```diff
@@ -36,22 +36,38 @@ do
...
     # arn:aws:lambda:{env.region}:017000801446:layer:AWSLambdaPowertoolsPython:39
     if [[ "$line" == *"eu-central-1"* ]]; then
-      for region in '{region}' '${AWS::Region}' '${aws:region}' '{env.region}'
+      # These are all the framework pseudo parameters currently found in the docs
+      for pseudo_region in '{region}' '${AWS::Region}' '${aws:region}' '{env.region}'
       do
-        generic_prefix=$(echo "$prefix" | sed "s/eu-central-1/${region}/")
-        generic_line=$(echo "$line" | sed "s/eu-central-1/${region}/")
+        generic_prefix=$(echo "$prefix" | sed "s/eu-central-1/${pseudo_region}/")
+        # generic_prefix = arn:aws:lambda:${AWS::Region}:017000801446:layer:AWSLambdaPowertoolsPython
```
Comment (heitorlessa) [inline layer/scripts/update_layer_arn.sh]:
```suggestion
        # prefix_pseudo_region = arn:aws:lambda:${AWS::Region}:017000801446:layer:AWSLambdaPowertoolsPython
```

File: layer/scripts/update_layer_arn.sh [hunk 12]
```diff
@@ -36,22 +36,38 @@ do
...
+
+        generic_line=$(echo "$line" | sed "s/eu-central-1/${pseudo_region}/")
+        # generic_line = arn:aws:lambda:${AWS::Region}:017000801446:layer:AWSLambdaPowertoolsPython:49
 
+        # Replace all the "generic_prefix"'s in the file
```
Comment (heitorlessa) [inline layer/scripts/update_layer_arn.sh]:
```suggestion
        # Replace all the "prefix_pseudo_region"'s in the file
```

**Review by heitorlessa — CHANGES_REQUESTED** [review]
looks great! added last suggestions on var names

**Comment by heitorlessa** [comment]
Consider it approved after these last changes to ease maintenance.

**Review by heitorlessa — APPROVED** [review]

**Pull request merged**